firestore/bulkwriter.go

```diff
@@ -0,0 +1,319 @@
+		return nil, fmt.Errorf("bulkwriter: early write cancellation")
+	case wpb := <-j.result:
+		return writeResultFromProto(wpb)
+	case err := <-j.err:
+		return nil, err
+	}
+}
+
+// setError ensures that an error is returned on the error channel of BulkWriterJob.
+func (j *BulkWriterJob) setError(e error) {
+	j.err <- e
+}
+
+// setSuccess ensures that a WriteResult is returned to the result channel of BulkWriterJob.
+func (j *BulkWriterJob) setResult(r *pb.WriteResult) {
+	j.result <- r
+}
+
+// A BulkWriter supports concurrent writes to multiple documents. The BulkWriter
+// submits document writes in maximum batches of 20 writes per request. Each
+// request can contain many different document writes: create, delete, update,
+// and set are all supported.
+//
+// Only one operation (create, set, update, delete) per document is allowed.
+// BulkWriter cannot promise atomicity: individual writes can fail or succeed
+// independent of each other. Bulkwriter does not apply writes in any set order;
+// thus a document can't have set on it immediately after creation.
+type BulkWriter struct {
+	database        string             // the database as resource name: projects/[PROJECT]/databases/[DATABASE]
+	start           time.Time          // when this BulkWriter was started; used to calculate qps and rate increases
+	vc              *vkit.Client       // internal client
+	maxOpsPerSecond int                // number of requests that can be sent per second
+	docUpdatePaths  map[string]bool    // document paths with corresponding writes in the queue
+	cancel          context.CancelFunc // context to send cancel message
+	bundler         *bundler.Bundler   // handle bundling up writes to Firestore
+	ctx             context.Context    // context for canceling all BulkWriter operations
+	openLock        sync.Mutex         // guards against setting isOpen concurrently
+	isOpen          bool               // flag that the BulkWriter is closed
+}
+
+// newBulkWriter creates a new instance of the BulkWriter. This
+// version of BulkWriter is intended to be used within go routines by the
+// callers.
+func newBulkWriter(ctx context.Context, c *Client, database string) *BulkWriter {
+	// Although typically we shouldn't store Context objects, in this case we
+	// need to pass this Context through to the Bundler
+	ctx, cancel := context.WithCancel(withResourceHeader(ctx, c.path()))
+
+	bw := &BulkWriter{
+		database:        database,
+		start:           time.Now(),
+		vc:              c.c,
+		isOpen:          true,
+		maxOpsPerSecond: defaultStartingMaximumOpsPerSecond,
+		docUpdatePaths:  make(map[string]bool),
+		cancel:          cancel,
+		ctx:             ctx,
+	}
+
+	// can't initialize with struct above; need instance reference to BulkWriter.send()
+	bw.bundler = bundler.NewBundler(BulkWriterJob{}, bw.send)
+	bw.bundler.HandlerLimit = bw.maxOpsPerSecond
+	bw.bundler.BundleCountThreshold = maxBatchSize
+	bw.bundler.BundleByteLimit = 0 // unlimited size
```

Comment:
Can be removed since 0 is the default value for this type.

Reply:
Done.